ansible_base/activitystream/models/entry.py

```diff
@@ -0,0 +1,81 @@
+from functools import partial
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.signals import m2m_changed, post_save, pre_delete, pre_save
+from django.utils.translation import gettext_lazy as _
+
+from ansible_base.activitystream.signals import activitystream_create, activitystream_delete, activitystream_update, activitystream_m2m_changed
+from ansible_base.lib.abstract_models import CommonModel
+
+
+class Entry(CommonModel):
+    """
+    An activity stream entry.
+
+    This is keyed on a generic object_id and content_type, which allows for
+    a wide variety of objects to be used in the activity stream.
+    """
+
+    class Meta:
+        verbose_name_plural = _('Entries')
+
+    OPERATION_CHOICES = [
+        ('create', _('Entity created')),
+        ('update', _("Entity updated")),
+        ('delete', _("Entity deleted")),
+        ('associate', _("Entity was associated with another entity")),
+        ('disassociate', _("Entity was disassociated with another entity")),
+    ]
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING)
+    object_id = models.TextField(null=True, blank=True)
+    content_object = GenericForeignKey('content_type', 'object_id')
+    operation = models.CharField(max_length=12, choices=OPERATION_CHOICES)
+    changes = models.JSONField(null=True, blank=True)
+
+    # This is used for m2m (dis)associations
+    related_content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING, null=True, blank=True, related_name='related_content_type')
+    related_object_id = models.TextField(null=True, blank=True)
+    related_content_object = GenericForeignKey('related_content_type', 'related_object_id')
+
+    # TODO: AWX stores denormalized actor data to account for deleted users, we should do the same
```

Comment:
Since you're using `DO_NOTHING`, I think it's actually a much better policy to leave it as it is and document what's possible. Get the user id, and then look at the activity stream filtered for the user type and id. If you want to really hold their hand, then you could add  "related" link to the activity stream for content_object and related_content_object, which should still be valid for deleted objects.

Something felt wrong about what AWX did, and there's legitimacy to that feeling. The deletion record for the user will show the username, and that will be saved according to the user's own retention policy. Saving a serialized version of related objects doesn't really obey the user's intent properly.